src/python/ensembl/io/genomio/gff3/id_allocator.py

```diff
@@ -0,0 +1,191 @@
+        return transcript_id
+
+    def normalize_cds_id(self, cds_id: str) -> str:
+        """Returns a normaliSed version of the provided CDS ID.
+
+        The normalisation implies to remove any unnecessary prefixes around the CDS ID. However, if
+        the CDS ID is still not proper, an empty string will be returned.
+
+        Args:
+            cds_id: CDS ID to normalize.
+
+        """
+
+        prefixes = ["cds-", "cds:"]
+        normalized_cds_id = StableIDAllocator.remove_prefix(cds_id, prefixes)
+
+        # Special case: if the ID doesn't look like one, remove it - it needs to be regenerated
+        if not self.is_valid(normalized_cds_id):
+            return ""
+        return normalized_cds_id
+
+    def normalize_pseudogene_cds_id(self, pseudogene: SeqFeature) -> None:
+        """Normalizes every CDS ID of the provided pseudogene.
+
+        Ensure each CDS from a pseudogene has a proper ID:
+        - Different from the gene
+        - Derived from the gene if it is not proper
+
+        Args:
+            pseudogene: Pseudogene feature.
+        """
+        for transcript in pseudogene.sub_features:
+            for feat in transcript.sub_features:
+                if feat.type == "CDS":
+                    feat.id = self.normalize_cds_id(feat.id)
+                    if feat.id in ("", pseudogene.id):
+                        feat.id = f"{transcript.id}_cds"
+                        feat.qualifiers["ID"] = feat.id
+
+    def normalize_gene_id(self, gene: SeqFeature) -> str:
+        """Returns a normalized gene stable ID.
+
+        Removes any unnecessary prefixes, but will generate a new stable ID if the normalized one is
+        not recognized as valid.
+
+        Args:
+            gene: Gene feature to normalize.
+        """
+        prefixes = ["gene-", "gene:"]
+        new_gene_id = StableIDAllocator.remove_prefix(gene.id, prefixes)
+
+        # In case the normalized gene ID is not valid, use the GeneID
+        if not self.is_valid(new_gene_id):
+            logging.warning(f"Gene ID is not valid: {new_gene_id}")
+            qual = gene.qualifiers
+            if "Dbxref" in qual:
+                for xref in qual["Dbxref"]:
+                    (db, value) = xref.split(":")
+                    if db != "GeneID":
+                        continue
+                    new_gene_id_base = f"{db}_{value}"
+                    new_gene_id = new_gene_id_base
+                    number = 1
+                    while new_gene_id in self._loaded_ids:
+                        number += 1
+                        new_gene_id = f"{new_gene_id_base}_{number}"
+                        if number > 10:
+                            raise InvalidStableID(f"Duplicate ID {new_gene_id_base} (up to {new_gene_id})")
+                    self._loaded_ids.add(new_gene_id)
+                    logging.debug(f"Using GeneID {new_gene_id} for stable_id instead of {gene.id}")
+                    return new_gene_id
```

Comment:
I'm struggling to understand the usage of `_loaded_ids`, in particular why you only add them here and not when a new `gene_id` is generated in the method above.